### ISSUE TITLE
Firefox Send - downloads fail

### ISSUE THREAD

**Issue by smaragdus** [erickutcher/httpdownloader#60]
Someone sent me some files uploaded to [Firefox Send](https://send.firefox.com/) and HTTP Downloader (version 1.0.2.2) failed to download these files showing error `Authorization Required`, screen:

![HTTP Downloader 1 0 2 2 - 2019-06-08 - auth - 003](https://user-images.githubusercontent.com/9025757/59135958-a0a1e480-8989-11e9-8968-ac15c14b1eb4.png)

The files I was trying to download via HTTP Downloader were not password-protected. However other download managers failed to get these files too. When I disabled HTTP Downloader I was able to get these files via the browser. I am not in the least a fan of Mozilla and its products and service, I am just curious why these downloads fail. I prepared some 1 KB text files for testing:

- [File 1](https://send.firefox.com/download/0378459e7733e68f/#sbMvV0GH5pDFg6qXuQdRxg)
- [File 2](https://send.firefox.com/download/9c8039fb8234e374/#w6aLMmNgDLVe5RUfGcwwgg)
- [File 3](https://send.firefox.com/download/a975e2dee85262b4/#oEd3NgX-Mgr0ky7uq3wTNA)
- [File 4](https://send.firefox.com/download/031a8a3b0110a3a6/#uWf-QEyA3pKRSSX9HfE10A)

I should mention that these download links will expire either after 1 download or after 24 hours- Mozilla does not allow me to change the settings without logging in and I refuse to give Mozilla any private data.

**Comment by erickutcher** [comment]
I'll have to get to this on the weekend. I've been swamped lately with other things. Sorry about that.

**Comment by smaragdus** [comment]
@erickutcher 

No problem at all, this is neither important nor urgent- I simply noticed that HTTP Downloader failed to get these files.

The download links I created have expired. When you have time you may upload a file using [Firefox Send](https://send.firefox.com/) and see why HTTP Downloader fails to get the file.

**Comment by erickutcher** [comment]
So I think the download page generates an authorization key that gets passed in the HTTP Authorization header. The problem with it is that it's a one time key and when you override the browser's download manager, it'll end up getting used before it gets sent to HTTP Downloader. The browser initializes the download and the extensions stop it before it goes further, but by that time the key will have been submitted and invalidated. At least I think that's what's happening.

If the key is being generated with JavaScript, then I don't think I'll be able to solve it. For some reason I can't get Firefox to prompt me on that site.

Needs to be looked into a little more.

**Comment by smaragdus** [comment]
@erickutcher 

Please feel free to close this issue. I do not use Firefox Send. It just happened that I came across this problem. I may close the issue myself but you might like to investigate further.

**Comment by Zero3K** [comment]
I think that it should be looked into since others might be using it.

**Comment by smaragdus** [comment]
Firefox Send has been discontinued:

- [Mozilla shuts down Firefox Send encrypted file share service due to malware/phishing](https://alternativeto.net/news/2020/9/mozilla-shuts-down-firefox-send-encrypted-file-share-service-due-to-malware-phishing/)
- [Update on Firefox Send and Firefox Notes](https://blog.mozilla.org/blog/2020/09/17/update-on-firefox-send-and-firefox-notes/)

So I am closing this issue.

I won't bet that Firefox and Mozilla will still exist in two or three years.

**Comment by erickutcher** [comment]
Seems like it was bound to happen. File lockers come and go.

**Comment by SeriousHoax** [comment]
@erickutcher Hi! I have the same Authorization error while downloading from, `https://cs.rin.ru/forum/`.
An example download link, `https://cs.rin.ru/forum/download/file.php?id=175180`
Source: `https://cs.rin.ru/forum/viewtopic.php?p=3415683&sid=382d5ce228793f2a0dbb753fc8a8992e#p3415683`
You would need an account to see the links on the site.

<img width="646" height="22" alt="Image" src="https://github.com/user-attachments/assets/dfe3231a-27e3-4482-8657-3905580c28a0" />

**Comment by erickutcher** [comment]
@SeriousHoax That download requires three things. It needs a User-Agent, Referer, and Cookie.

You'll want to use one of the browser extensions in order to get that information.

In the Add URL(s) window you can put a check in "Advanced options" and add this to the "Headers" tab:

```
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:145.0) Gecko/20100101 Firefox/145.0
Referer: https://cs.rin.ru/forum/viewtopic.php?p=3415683&sid=382d5ce228793f2a0dbb753fc8a8992e#p3415683
```

The "Cookies" tab needs to be filled out with whatever cookies have been saved for that site. It's best to use the browser extension to get this information.

**Comment by SeriousHoax** [comment]
@erickutcher Hi! I'm already using the browser extension. I always use it. I also tried with and without boringssl, just in case. I still get the error message.

<details>

![Image](https://github.com/user-attachments/assets/2587051b-85b8-4e46-99fa-bd09ab5e6d7a)

</details>

**Comment by erickutcher** [comment]
Click on the Advanced options when the Add URL(s) window pops up and make sure that the Cookies tab is filled out and the Headers tab a well.

I tried it with Firefox and Edge and it worked fine. You might try clearing your cache and cookies and then signing back into the site. Maybe it's a cookie issue.

**Comment by SeriousHoax** [comment]
Deleted cache and cookies, logged in yet have the same issue on both Edge and Firefox. Cookies and Headers seem fine also. What could be the issue? 

<details>

<img width="582" height="531" alt="Image" src="https://github.com/user-attachments/assets/f1116900-cd9f-4c70-bb04-bf72b0a9d0d9" />
<img width="466" height="283" alt="Image" src="https://github.com/user-attachments/assets/8fce63aa-6e79-4d42-9f4b-50c88b386c90" />
<img width="563" height="285" alt="Image" src="https://github.com/user-attachments/assets/f3278094-8eea-4e98-9cf4-0732cd183a4d" />

</details>

**Comment by erickutcher** [comment]
Check the Site Manager and make sure there's no settings that are interfering with the cs.rin.ru domain. Also check the Update Download window (right click on the download in the list and select Update Download...) and make sure that the Cookies and Headers match what you see in the Add URL(s) window.

I'm able to download files from that site so I'm not sure what the issue could be.

**Comment by SeriousHoax** [comment]
My Site Manager is empty. Weird that it's working for you but not me. Also works with Free Download Manager.
The cookies and headers seem to be the same in advanced option and update url, though the cookie's u and k values are first on one but last on another. Maybe this is expected? I don't know. 

<details>

From, Advanced Option:

csrinru_phpbb3_u=1607142; csrinru_phpbb3_k=236d00e2494f0fe6; csrinru_phpbb3_sid=4a974fa9ea221f8b310571272c5b8b55; securitytoken=1TObw7_oOs7BFkBnKKK2uw; securitytoken_expiration=1766055101

User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/143.0.0.0 Safari/537.36 Edg/143.0.0.0
Referer: https://cs.rin.ru/forum/viewtopic.php?p=3415683

From Update URL:

csrinru_phpbb3_k=236d00e2494f0fe6; csrinru_phpbb3_sid=4a974fa9ea221f8b310571272c5b8b55; csrinru_phpbb3_u=1607142; securitytoken=1TObw7_oOs7BFkBnKKK2uw; securitytoken_expiration=1766055101

User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/143.0.0.0 Safari/537.36 Edg/143.0.0.0
Referer: https://cs.rin.ru/forum/viewtopic.php?p=3415683

</details>

This is my portable HTTP Download Manager. You may check this if required to see if anything is wrong with any settings or not and test this on that website.

https://www.upload.ee/files/18902598/HTTP_Downloader.zip.html

**Comment by erickutcher** [comment]
Your settings look fine and I was able to download using your setup without issue.

I did some testing with the user-agent string and noticed that if it's not the exact same string as what you used when logging into that site, then it's most likely not going to work. I installed a user-agent switcher/changer, changed it to something other than what the browser uses, logged into the site, downloaded the file, and it failed with the Authorization Required status. When I updated the user-agent to what the browser's default user-agent is, then it worked.

Basically if you have a user-agent switcher/changer extension enabled, then you'll want to disable it or copy the user-agent string that it uses and enter it into the Site Manager for the cs.rin.ru domain.

**Comment by SeriousHoax** [comment]
Such a strange case. 
I don't use any user-agent switcher extension. 
You can see my installed extension in this GIF that I previously shared: 

https://github.com/erickutcher/httpdownloader/issues/60#issuecomment-3657359706

So I don't know what's wrong. FDM didn't have any issue downloading. Wouldn't that mean that cookies and user agent are ok? 
Is there any way for me to capture logs and send it to you?

**Comment by erickutcher** [comment]
It's probably one of the Violentmonkey scripts that's being loaded. I tested AdGuard AdBlocker and it didn't affect things.

If the user-agent is being changed, then you can confirm it by going here: https://whatmyuseragent.com/

Try temporarily disabling all of the extensions except for HTTP Downloader. Log out of the site or clear your cookies, and then log back into it to download the file.

I looked into a different way in which to retrieve the user-agent and if what I've described above is what's happening, then I think I can get a fix for the issue.

**Comment by SeriousHoax** [comment]
Two userscripts are injected on that page. On is AdGuard Extra: https://github.com/AdguardTeam/AdGuardExtra
Another one is created by me with the help of Claude: https://raw.githubusercontent.com/SeriousHoax/AnythingAdblock/refs/heads/master/Userscripts/cs.rin.ru-Remove-Unnecessary-Images.js

They don't change the user-agent. But even without them it's the same. 
On the site it shows fine also:

<img width="734" height="577" alt="Image" src="https://github.com/user-attachments/assets/aca52273-2b7d-4577-b4a5-89dceeca5be5" />

**Comment by erickutcher** [comment]
I'm stumped. What happens if you sign up for a new account and try the download from that? Maybe there's someone else that could try it on their system.
